homeassistant/components/tado/sensor.py

```diff
@@ -149,97 +144,138 @@ def icon(self):
+            self._state_attributes = {
+                "time": self._tado_zone_data.heating_power_timestamp
+            }
 
         elif self.zone_variable == "ac":
-            if "activityDataPoints" in data:
-                activity_data = data["activityDataPoints"]
-
-                if "acPower" in activity_data and activity_data["acPower"] is not None:
-                    self._state = activity_data["acPower"]["value"]
-                    self._state_attributes = {
-                        "time": activity_data["acPower"]["timestamp"]
-                    }
+            self._state = self._tado_zone_data.ac_power
+            self._state_attributes = {"time": self._tado_zone_data.ac_power_timestamp}
 
         elif self.zone_variable == "tado bridge status":
-            if "connectionState" in data:
-                self._state = data["connectionState"]["value"]
+            self._state = self._tado_zone_data.connection
 
         elif self.zone_variable == "tado mode":
-            if "tadoMode" in data:
-                self._state = data["tadoMode"]
+            self._state = self._tado_zone_data.tado_mode
 
         elif self.zone_variable == "overlay":
-            self._state = "overlay" in data and data["overlay"] is not None
+            self._state = self._tado_zone_data.overlay_active
             self._state_attributes = (
-                {"termination": data["overlay"]["termination"]["type"]}
-                if self._state
+                {"termination": self._tado_zone_data.overlay_termination_type}
+                if self._tado_zone_data.overlay_active
                 else {}
             )
 
         elif self.zone_variable == "early start":
-            self._state = "preparation" in data and data["preparation"] is not None
+            self._state = self._tado_zone_data.preparation
 
         elif self.zone_variable == "open window":
-            self._state = "openWindow" in data and data["openWindow"] is not None
-            self._state_attributes = data["openWindow"] if self._state else {}
+            self._state = self._tado_zone_data.open_window
+            self._state_attributes = self._tado_zone_data.open_window_attr
+
+
+class TadoDeviceSensor(Entity):
+    """Representation of a tado Sensor."""
+
+    def __init__(self, hass, tado, device_name, device_id, device_variable):
+        """Initialize of the Tado Sensor."""
+        self.hass = hass
+        self._tado = tado
+
+        self.device_name = device_name
+        self.device_id = device_id
+        self.device_variable = device_variable
+
+        self._unique_id = f"{device_variable} {device_id} {tado.device_id}"
+
+        self._state = None
+        self._state_attributes = None
+        self._tado_device_data = None
+        self._async_update_device_data()
+
+    async def async_added_to_hass(self):
+        """Register for sensor updates."""
+
+        async_dispatcher_connect(
```

Comment:
Don't forget to unsubscribe in async_will_remove_from_hass

Reply:
Good catch.  When I broke apart the sensor I missed that.  Its missing in all places. Added